File: Anul 2/Metode Avansate de Programare/Seminarii/Seminar 1/src/runner/AbstractTaskRunner.java
package runner;

import domain.Task;

public abstract class AbstractTaskRunner implements TaskRunner
{
    private TaskRunner runner;

    public AbstractTaskRunner(TaskRunner runner)
    {
        this.runner = runner;
    }

    @Override
    public void executeOneTask()
    {
        runner.executeOneTask();
    }

    @Override
    public void executeAll()
    {
        while  (hasTask())
        {
            executeOneTask();
        }
    }

    @Override
    public void addTask(Task t)
    {
        runner.addTask(t);
    }

    @Override
    public boolean hasTask()
    {
        return runner.hasTask();
    }
}

File: Anul 2/Metode Avansate de Programare/Seminarii/sem11-taskrunner/taskrunner/taskrunner/runner/AbstractTaskRunner.cs
namespace taskrunner.runner;

public abstract class AbstractTaskRunner : ITaskRunner
{
    protected ITaskRunner runner;

    protected AbstractTaskRunner(ITaskRunner runner)
    {
        this.runner = runner;
    }

    public virtual void ExecuteOneTask()
    {
        runner.ExecuteOneTask();
    }

    // dacă vrei o implementare diferită, suprascrii metoda în subclasă
    // public abstract void ExecuteOneTask();

    public virtual void ExecuteAll()
    {
        while (runner.HasTask())
        {
            ExecuteOneTask();
        }
    }

    public virtual void AddTask(Task t)
    {
        runner.AddTask(t);
    }

    public virtual bool HasTask()
    {
        return runner.HasTask();
    }
}